lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -2080,6 +2086,38 @@ static LogicalResult checkConnectFlow(Operation *connect) {
     return diag.attachNote(dstRef.getLoc())
            << "the destination was defined here";
   }
+
+  // If the connect is for RefType, implement the constraint for downward only
+  // references. We cannot connect
+  //  1. an input reference port to the output reference port.
+  //  2. instance reference results to each other.
+  //  This means, the connect can only be used for forwarding RefType module
+  //  ports to Instance ports.
+  if (dst.getType().isa<RefType>()) {
+    if (getDeclarationKind(src) == getDeclarationKind(dst)) {
```

Comment:
It seems weird to include this ref-specific check inside of flow checking. Is there a better home for this inside connect checking?